anabot/runtime/installation/hub/partitioning/specialized_disks/__init__.py

```diff
@@ -0,0 +1,73 @@
+# -*- coding: utf-8 -*-
+
+import logging
+logger = logging.getLogger('anabot')
+
+from fnmatch import fnmatchcase
+
+from anabot.runtime.decorators import handle_action, handle_check, check_action_result
+from anabot.runtime.default import default_handler
+from anabot.runtime.functions import get_attr, getnode, getnodes, getsibling
+from anabot.runtime.errors import TimeoutError
+from anabot.runtime.translate import tr
+from anabot.runtime.installation.common import done_handler
+from anabot.runtime.actionresult import NotFoundResult as NotFound, ActionResultPass as Pass, ActionResultFail as Fail
+
+_local_path = '/installation/hub/partitioning/add_spec_disk'
+def handle_act(path, *args, **kwargs):
+    return handle_action(_local_path + path, *args, **kwargs)
+
+def handle_chck(path, *args, **kwargs):
+    return handle_check(_local_path + path, *args, **kwargs)
+
+def select_manipulate(element, app_node, local_node, dryrun):
+    name = get_attr(element, "names", "*")
+    action = get_attr(element, "action", "select")
+
+    try:
+        results_label = getnode(app_node, "label", tr("Search Res_ults:", context="GUI|Installation Destination|Filter|Search"))
+        table_pane = getsibling(results_label, 1, "scroll pane")
+        disks_table = getnode(table_pane, "table")
+    except TimeoutError as e:
+        return NotFound("Results found label, table pane or disks table: %s" % e)
+    cells = getnodes(disks_table, "table cell")
+    # every table row has 12 cells, with checkbox on position 0 and device name on 1
+    name_cells = [
+                    getsibling(c, -1, "table cell") 
+                    for c in cells[1::12]
+                    if fnmatchcase(c.name, name)
+    ]
+    logger.info("Devices that match name '%s' found: '%s'" % (name, [n.name for n in name_cells]))
+    for c in name_cells:
+        checkbox = getsibling(c, -1, "table cell")
+        if (action == "select") != (c.checked):
+            if dryrun:
+                return Fail("Checkbox status for device '%s' doesn't match required action '%s'" % (c.name, action))
+            logger.info("Clicking on checkbox for device %s" % c.name)
+            c.click()
+    return Pass()
+
+@handle_act('/select')
+def select_handler(element, app_node, local_node):
+    return select_manipulate(element, app_node, local_node, False)
+
+@handle_chck('/select')
+def select_check(element, app_node, local_node):
+    return select_manipulate(element, app_node, local_node, True)
+
+@handle_act('/done')
+def done_manipulate(element, app_node, local_node):
+    try:
+        done_button = getnode(app_node, "push button", tr("_Done", drop_underscore=False, context="GUI|Spoke Navigation"))
+    except TimeoutError:
+        return NotFound("clickable Done button")
+    done_button.click()
+    return Pass()
+
+@handle_chck('/done')
+def done_check(element, app_node, local_node):
```

Comment:
This check seems to have reversed logic. Let's say there's <done policy="just_check"/> testing if the done button is available and clickable, this check implementation would fail. This check looks like a check for /installation/hub/partitioning/add_spec_disk

There's a decorator @check_action_result that could be used here like in /installation/hub/partitioning/advanced/done

Reply:
Unless I missed your point I think the code is OK. The logic behind it is that after clicking on 'Done' in the Specialized & network disks spoke, this spoke should disappear and the basic partitioning spoke should be visible. This can be easily achieved by looking up a visible 'Add Specialized Disk' button, although it might deserve some comment to make it more obvious.

Reply:
Let's say I have following Anabot recipe:
```xml
<ez:installation>
  <ez:hub>
    <partitioning>
      <add_spec_disk>
        <done policy="just_check" />
        <done />
      </add_spec_disk>
    </partitioning>
  </ez:hub>
</ez:installation>
```
Such recipe would fail at the `<done policy="just_check" />` because it wouldn't find "Add Specialized Disk" push button.

Reply:
I'm still not quite convinced this behaviour is wrong. Let me explain: 1. You enter the specialized & network disks subspoke. 2. You want to check that an action performed by clicking on Done button within the subspoke has been performed successfully. The check fails as the checked action wasn't performed - the button wasn't clicked on, therefore nothing changed. You're still in the subspoke, not back in the main partitioning spoke as you would be if the Done button was used. Therefore I think the fail is correct in this case.

On the contrary, if you click on the Done button in the subspoke, you should get back to the main partitioning spoke, the check would make sure the required button is visible and will Pass if it really is, or Fail if it's not (if something failed and you got stuck in the subspoke for some reason).

Reply:
So the failure of `<done policy="just_check" />` (/installation/hub/partitioning/add_specialized_disk/done) should always fail and it's expected?

The way how just_check is used on other places (or more importantly how it should be used is) that it only checks not performing any action. In this case, the just_check should be successful if the Done button is available and can be used.

@jstodola what are your thoughts here?

Reply:
I just realized now that also `<done policy="just_check" />` (/installation/hub/partitioning/add_specialized_disk/done) would actually pass.

Reply:
> So the failure of `<done policy="just_check" />` (/installation/hub/partitioning/add_specialized_disk/done) should always fail and it's expected?

I'd say it's expected, since the action that `/installation/hub/partitioning/add_specialized_disk/done` handler should arrange for didn't happen in such a case, therefore the check should fail. And I think that using the check with `just_check` actually doesn't make much sense there.

As for `<done policy="just_check_fail" />`, I'm not sure what you exactly meant here by _pass_ - did you mean the result of the check itself, or of the final check result with the policy taken into account? The way it is implemented results in a failed check for `just_check` and expected failure for `just_check_fail`, which I think is OK and is consistent with what I mentioned before.

Reply:
@jstodola, I'm not sure what you mean by that it shouldn't do any action. Do you mean the done handler shouldn't be implemented at all and the button should be clicked automatically when the processing of add_specialized_disks handler finishes processing of all of its children? Or something else?

Reply:
We've discussed this face-to-face and came to agreement about the solution. Resolving this conversation as the agreed solution is implemented now.